src/compiler/checker.ts

```diff
@@ -10428,12 +10433,37 @@ namespace ts {
             return reduced;
         }
 
+        function isNeverReducedProperty(prop: Symbol) {
+            return isDiscriminantWithNeverType(prop) || isConflictingPrivateProperty(prop);
+        }
+
         function isDiscriminantWithNeverType(prop: Symbol) {
+            // Return true for a synthetic non-optional property with non-uniform types, where at least one is
+            // a literal type and none is never, that reduces to never.
             return !(prop.flags & SymbolFlags.Optional) &&
                 (getCheckFlags(prop) & (CheckFlags.Discriminant | CheckFlags.HasNeverType)) === CheckFlags.Discriminant &&
                 !!(getTypeOfSymbol(prop).flags & TypeFlags.Never);
         }
 
+        function isConflictingPrivateProperty(prop: Symbol) {
+            // Return true for a synthetic property with multiple declarations, at least one of which is private.
+            return !prop.valueDeclaration && !!(getCheckFlags(prop) & CheckFlags.ContainsPrivate);
+        }
+
+        function elaborateNeverIntersection(errorInfo: DiagnosticMessageChain | undefined, type: IntersectionType) {
+            const neverProp = find(getPropertiesOfUnionOrIntersectionType(type), isDiscriminantWithNeverType);
+            if (neverProp) {
+                return chainDiagnosticMessages(errorInfo, Diagnostics.The_intersection_0_was_reduced_to_never_because_property_1_has_conflicting_types_in_some_constituents,
+                    typeToString(type, /*enclosingDeclaration*/ undefined, TypeFormatFlags.NoTypeReduction), symbolToString(neverProp));
+            }
+            const privateProp = find(getPropertiesOfUnionOrIntersectionType(type), isConflictingPrivateProperty);
+            if (privateProp) {
+                return chainDiagnosticMessages(errorInfo, Diagnostics.The_intersection_0_was_reduced_to_never_because_property_1_exists_in_multiple_constituents_and_is_private_in_some,
+                    typeToString(type, /*enclosingDeclaration*/ undefined, TypeFormatFlags.NoTypeReduction), symbolToString(privateProp));
+            }
+            return errorInfo;
+        }

```

Comment:
Consider reporting more than the first conflict

```suggestion
        function tryElaborateIfIntersectionReducedToNever(type: Type, currentErrorInfo: DiagnosticMessageChain | undefined) {
            if (getObjectFlags(type) & ObjectFlags.IsNeverIntersection) {
                const intersectionString = typeToString(type, /*enclosingDeclaration*/ undefined, TypeFormatFlags.PreserveVacuousIntersections);
                const conflictingPropertyNames = map(filter(getPropertiesOfUnionOrIntersectionType(<IntersectionType>type), isDiscriminantWithNeverType), symbolName);
                const maxNumProperties = 3;
                const propertiesToReport = conflictingPropertyNames.length <= maxNumProperties
                    ? conflictingPropertyNames
                    : [...conflictingPropertyNames.slice(0, maxNumProperties), "..."];
                const propertiesString = propertiesToReport.join(", ");
                const diag = conflictingPropertyNames.length === 1
                    ? Diagnostics.The_type_never_was_reduced_from_the_intersection_0_Some_types_of_that_intersection_conflict_in_the_property_1_so_values_of_that_type_can_never_exist
                    : Diagnostics.The_type_never_was_reduced_from_the_intersection_0_Some_types_of_that_intersection_conflict_in_the_properties_1_so_values_of_that_type_can_never_exist;
                return chainDiagnosticMessages(currentErrorInfo, diag, intersectionString, propertiesString);
            }
            return currentErrorInfo;
        }
```

Reply:
I looked at doing that but decided it wasn't worth the extra code and error messages. You feel strongly we need it?